### ISSUE TITLE
Caret versions don't resolve properly when using custom installer path

### ISSUE THREAD

**Issue by SamMousa** [francoispluchino/composer-asset-plugin#172]
Is the `^` properly supported in version constraints?

When doing for example `composer show bower-asset/jquery` I get:
```` 3.0.0-alpha1, 2.1.4, 2.1.3, 2.1.2, 2.1.1, 2.1.1-RC2, 2.1.1-RC1, 2.1.1-beta1, 2.1.0, 2.1.0-RC1, 2.1.0-beta3, 2.1.0-beta2, 2.1.0-beta1, 2.0.3, 2.0.2, 2.0.1, 2.0.0, 2.0.0-beta3, 2.0.0-beta2, 2.0.0-beta1, 1.11.3, 1.11.2, 1.11.1, 1.11.1-RC2, 1.11.1-RC1, 1.11.1-beta1, 1.11.0, 1.11.0-RC1, 1.11.0-beta3, 1.11.0-beta2, 1.11.0-beta1, 1.10.2, 1.10.1, 1.10.0, 1.10.0-beta1, 1.9.1, 1.9.0, 1.9.0-RC1, 1.9.0-beta1, 1.8.3, 1.8.2, 1.8.1, 1.8.0, 1.8-RC1, 1.8-beta2, 1.8-beta1, 1.7.2, 1.7.2-RC1, 1.7.2-beta1, 1.7.1, 1.7.1-RC1, 1.7.0, 1.7-RC2, 1.7-RC1, 1.7-beta2, 1.7-beta1, 1.6.4, 1.6.4-RC1, 1.6.3, 1.6.3-RC1, 1.6.2, 1.6.2-RC1, 1.6.1, 1.6.1-RC1, 1.6.0, 1.6-RC1, 1.6-beta1, 1.5.2, 1.5.2-RC1, 1.5.1, 1.5.1-RC1, 1.5.0, 1.5-RC1, 1.5-beta1, 1.4.4, 1.4.4-RC3, 1.4.4-RC2, 1.4.4-RC1, 1.4.3, 1.4.3-RC2, 1.4.3-RC1, 1.4.2, 1.4.1, 1.4.0, 1.4-RC1, 1.4-alpha2, 1.4-alpha1, 1.3.2, 1.3.1, 1.3.1-RC1, 1.3.0, 1.3-RC1, 1.3-beta2, 1.3-beta1, 1.2.6, 1.2.5, 1.2.4, 1.2.4-beta1, 1.2.4-alpha1, 1.2.3, 1.2.3-beta1, 1.2.3-alpha1, 1.2.2, 1.2.2-beta2, 1.2.2-beta1, 1.2.1, 1.2, 1.1.4, 1.1.3.1, 1.1.3, 1.1.3-alpha1, 1.1.2, 1.1.1, 1.1, 1.1-beta1, 1.1-alpha1, 1.0.4, 1.0.3, 1.0.2, 1.0.1, 1.0, 1.0-alpha1, dev-2.2-stable, dev-1.12-stable, dev-killphp, dev-compat````

When doing `composer require bower-asset/jquery ^2.1.4` I get non-resolvable error from composer.

**Comment by cebe** [comment]
works for me in an empty directory:

```
$ composer require bower-asset/jquery "^2.1.4"
You are running composer with xdebug enabled. This has a major impact on runtime performance. See https://getcomposer.org/xdebug
./composer.json has been created
Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Installing bower-asset/jquery (2.1.4)       
    Loading from cache

Writing lock file
Generating autoload files
```

**Comment by SamMousa** [comment]
Okay I can confirm the basic case works, there seems to be problems when using a non standard component path. I'll show what I can below.

`composer show bower-asset/moment`:
````
name     : bower-asset/moment
descrip. :                                           
keywords : 
versions : dev-master, 2.10.6, 2.10.5, 2.10.3, 2.10.2, 2.9.0, 2.8.4, 2.8.3, 2.8.2, 2.8.1, 2.8.0, 2.7.0, 2.6.0, 2.5.1, 2.5.0, 2.4.0, 2.3.1, 2.3.0, 2.2.1, 2.2.0, 2.1.0, 2.0.0, 1.7.2, 1.7.1, 1.7.0, 1.6.2, 1.6.1, 1.6.0, 1.5.1, 1.5.0, 1.4.0, 1.3.0, 1.2.0, 1.1.2, 1.1.1, 1.1.0, 1.0.1, 0.6.1, 0.6.0, 0.5.2, 0.5.1, 0.5.0, 0.4.1, 0.4.0, 0.3.2, 0.3.1, 0.3.0, dev-develop, dev-feature/es6-modules
type     : bower-asset-library
source   : [git] https://github.com/moment/moment.git f3fbef9d9875bbff340b527dbe3f1c447a942f69
dist     : [zip] https://api.github.com/repos/moment/moment/zipball/f3fbef9d9875bbff340b527dbe3f1c447a942f69 f3fbef9d9875bbff340b527dbe3f1c447a942f69
names    : bower-asset/moment
````
Composer.json looks like this:
````
{
    "name": "root/test",
    "type": "library",
    "license": "MTI",
    "authors": [
        {
            "name": "Sam Mousa",
            "email": "sam@mousa.nl"
        }
    ],
    "minimum-stability": "dev",
    "require": {
        "sam-it/yii1-datatables": "^1.0"
    },
     "extra" : {
        "asset-installer-paths": {
            "bower-asset-library" : "application/components"
        }
    }
}
````
Running `composer update` then gives me:
````
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - sam-it/yii1-datatables v1.0.2 requires bower-asset/moment ^2.0 -> no matching package found.
    - sam-it/yii1-datatables v1.0.1 requires bower-asset/moment ^2.0 -> no matching package found.
    - sam-it/yii1-datatables v1.0.0 requires bower-asset/moment ^2.0 -> no matching package found.
    - sam-it/yii1-datatables v1.0.2 requires bower-asset/moment ^2.0 -> no matching package found.
    - Installation request for sam-it/yii1-datatables ^1.0 -> satisfiable by sam-it/yii1-datatables[v1.0.0, v1.0.1, v1.0.2].

Potential causes:
 - A typo in the package name
 - The package is not available in a stable-enough version according to your minimum-stability setting
   see <https://groups.google.com/d/topic/composer-dev/_g3ASeIFlrc/discussion> for more details.

Read <https://getcomposer.org/doc/articles/troubleshooting.md> for further common problems.
````
Removing the `asset-installer-paths` setting and running `composer update` again succeeds.

**Comment by SamMousa** [comment]
Just for completeness, this is the composer.json from `sam-it/yii1-datatables`:
````
{
	"name": "sam-it/yii1-datatables",
	"type": "library",
    "description": "Yii widget wrapper for DataTables",
    "license": "MIT",
	"minimum-stability" : "dev",
    "authors": [
        {
		    "name": "Sam Mousa",
            "email": "sam@mousa.nl"
        }
    ],
	"require" : {
		"php" : ">=5.4.0",
        "bower-asset/datatables" : "^1.10.0",
        "bower-asset/moment": "^2.0",
        "bower-asset/drmonty-datatables-plugins": "^1.0"
	},
	"autoload" : {
		"psr-4" : {
			"SamIT\\Yii1\\DataTables\\" : "src/"
		}
	}
}
````

**Comment by francoispluchino** [comment]
Fixed in `1.2.x-dev` version.
